precompile/contract.go

```diff
@@ -14,13 +24,87 @@ type PrecompileAccessibleState interface {
+	Run(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error)
 	// RequiredGas returns the amount of gas required to execute this precompile on [input].
 	RequiredGas(input []byte) uint64
 }
+
+// statefulPrecompileFunction defines a function implemented by a stateful precompile
+type statefulPrecompileFunction struct {
+	// selector is the function selector for this function
+	selector []byte
+	// execute is performed when this function is selected
+	execute RunStatefulPrecompileFunc
+	// requiredGas calculates the amount of gas consumed by this function on [input].
+	requiredGas func(input []byte) uint64
+}
+
+// newStatefulPrecompileFunction creates a stateful precompile function with the given arguments
+func newStatefulPrecompileFunction(selector []byte, execute RunStatefulPrecompileFunc, requiredGas func(input []byte) uint64) *statefulPrecompileFunction {
+	return &statefulPrecompileFunction{
+		selector:    selector,
+		execute:     execute,
+		requiredGas: requiredGas,
+	}
+}
+
+// statefulPrecompileWithFunctionSelectors implements StatefulPrecompiledContract by using 4 byte function selectors to pass
+// off responsibilities to internal execution functions.
+// Note: because we only ever read from [functions] there no lock is required to make it thread-safe.
+type statefulPrecompileWithFunctionSelectors struct {
+	functions map[string]*statefulPrecompileFunction
+}
+
+// newStatefulPrecompileWithFunctionSelectors generates new StatefulPrecompile using [functions] as the available functions.
+func newStatefulPrecompileWithFunctionSelectors(functions ...*statefulPrecompileFunction) StatefulPrecompiledContract {
+	contract := &statefulPrecompileWithFunctionSelectors{functions: make(map[string]*statefulPrecompileFunction)}
+	for _, function := range functions {
+		_, exists := contract.functions[string(function.selector)]
+		if exists {
+			panic(fmt.Errorf("cannot create stateful precompile with duplicated function selector: %q", function.selector))
+		}
+		contract.functions[string(function.selector)] = function
+	}
+
+	return contract
+}
+
+// Run selects the function using the 4 byte function selector at the start of the input and executes the underlying function on the
+// given arguments.
+func (s *statefulPrecompileWithFunctionSelectors) Run(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if len(input) < selectorLen {
+		return nil, suppliedGas, fmt.Errorf("missing function selector to precompile - input length (%d)", len(input))
+	}
+
+	selector := input[:selectorLen]
+	functionInput := input[selectorLen:]
+	function, ok := s.functions[string(selector)]
```

Comment:
nit: add an optional fallback function if there is no function selector and no arguments.